File: 0001-1000-easy-cpp/0232-implement-queue-using-stacks.cpp
// https://leetcode.com/problems/implement-queue-using-stacks
class MyQueue {
private:
    std::stack<int> stackTopDown;
    std::stack<int> stackBottomUp;
    int front;

public:
    MyQueue() {
    }

    // O(n)
    void push(int x) {
        while (this->stackTopDown.size() > 0) {
            this->stackBottomUp.push(this->stackTopDown.top());
            this->stackTopDown.pop();
        }

        if (this->stackBottomUp.empty()) {
            this->front = x;
        }
        this->stackBottomUp.push(x);
    }

    // O(n)
    int pop() {
        while (this->stackBottomUp.size() > 0) {
            this->stackTopDown.push(this->stackBottomUp.top());
            this->stackBottomUp.pop();
        }

        int x = this->stackTopDown.top();
        this->stackTopDown.pop();
        if (this->stackTopDown.size() > 0) {
            this->front = this->stackTopDown.top();
        }
        return x;
    }

    // O(1)
    int peek() {
        return this->front;
    }

    // O(1)
    bool empty() {
        return this->stackTopDown.empty() && this->stackBottomUp.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: 0001-1000-easy-python/0232-implement-queue-using-stacks.py
# https://leetcode.com/problems/implement-queue-using-stacks
from collections import deque

class MyQueue(object):

    def __init__(self):
        self.stackTopDown = deque()
        self.stackBottomUp = deque()
        self.front = None

    # O(n)
    def push(self, x: int) -> None:
        while (len(self.stackTopDown) > 0):
            self.stackBottomUp.append(self.stackTopDown.pop());

        if (len(self.stackBottomUp) == 0):
            self.front = x
        self.stackBottomUp.append(x)

    # O(n)
    def pop(self) -> int:
        while (len(self.stackBottomUp) > 0):
            self.stackTopDown.append(self.stackBottomUp.pop())

        x = self.stackTopDown.pop()
        if (len(self.stackTopDown) > 0):
            self.front = self.stackTopDown.pop()
            self.stackTopDown.append(self.front)
        return x

    # O(1)
    def peek(self) -> int:
        return self.front

    # O(1)
    def empty(self) -> bool:
        return (len(self.stackTopDown) == 0) and (len(self.stackBottomUp) == 0)

# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

